packages/std/src/timestamp.rs

```diff
@@ -1,18 +1,71 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use crate::math::Uint64;
+
 /// A point in time in nanosecond precision.
 ///
-/// This type cannot represent any time before the UNIX epoch because both fields are unsigned.
-pub struct Timestamp {
-    /// Absolute time in seconds since the UNIX epoch (00:00:00 on 1970-01-01 UTC).
-    pub seconds: u64,
-    /// The fractional part time in nanoseconds since `time` (0 to 999999999).
-    pub nanos: u64,
-}
+/// This type can represent times from 1970-01-01T00:00:00Z to 2554-07-21T23:34:33Z.
+#[derive(
+    Serialize, Deserialize, Copy, Clone, Default, Debug, PartialEq, Eq, PartialOrd, Ord, JsonSchema,
+)]
+pub struct Timestamp(Uint64);
 
 impl Timestamp {
-    pub fn plus_seconds(&self, addition: u64) -> Timestamp {
-        Timestamp {
-            seconds: self.seconds + addition,
-            nanos: self.nanos,
-        }
+    /// Creates a timestamp from nanoseconds since epoch
```

Comment:
Hmmm... no 2 value constructor?

We must do eg. `Timestamp::from_seconds(env.block.time).plus_nanos(env.block.nanos)`?

It is a bit verbose, but it also means you won't mix up the two (identically typed) arguments. I guess it is good like that

Reply:
The question is where you get a two component time stamp from and of that case really matters. The example you provided is covered by `BlockInfo.timestamp()` where the calculation is a bit raw but more direct. Otherwise seconds plus nanos works (even for `const`s!).